src/ooxml/java/org/apache/poi/xssf/streaming/SheetDataWriter.java

```diff
@@ -354,88 +355,54 @@ protected void outputQuotedString(String s) throws IOException {
             return;
         }
 
-        char[] chars = s.toCharArray();
-        int last = 0;
-        int length = s.length();
-        for (int counter = 0; counter < length; counter++) {
-            char c = chars[counter];
-            switch (c) {
-                case '<':
-                    writeLastChars(_out, chars, last, counter);
-                    last = counter + 1;
+        final String xa0 = Character.toString((char)0xa0);
+
+        for (Iterator<String> iter = new StringCodepointsIterable(s).iterator(); iter.hasNext();) {
+            String codepoint = iter.next();
+            switch (codepoint) {
+                case "<":
                     _out.write("&lt;");
                     break;
-                case '>':
-                    writeLastChars(_out, chars, last, counter);
-                    last = counter + 1;
+                case ">":
                     _out.write("&gt;");
                     break;
-                case '&':
-                    writeLastChars(_out, chars, last, counter);
-                    last = counter + 1;
+                case "&":
                     _out.write("&amp;");
                     break;
-                case '"':
-                    writeLastChars(_out, chars, last, counter);
-                    last = counter + 1;
+                case "\"":
                     _out.write("&quot;");
                     break;
                 // Special characters
-                case '\n':
-                    writeLastChars(_out, chars, last, counter);
+                case "\n":
                     _out.write("&#xa;");
-                    last = counter + 1;
                     break;
-                case '\r':
-                    writeLastChars(_out, chars, last, counter);
+                case "\r":
                     _out.write("&#xd;");
-                    last = counter + 1;
                     break;
-                case '\t':
-                    writeLastChars(_out, chars, last, counter);
+                case "\t":
                     _out.write("&#x9;");
-                    last = counter + 1;
                     break;
-                case 0xa0:
-                    writeLastChars(_out, chars, last, counter);
+                case "\u00A0":
```

Comment:
Should we escape all Unicode codepoints to the equivalent html entity?

Otherwise, can you use the nbsp variable from above?

Reply:
Ideally, we want to escape as few chars as possible for readability